pkg/resource/db_parameter_group/hooks.go

```diff
@@ -214,3 +252,392 @@ func equalStrings(a, b *string) bool {
+	}
+	return familyMeta, nil
+}
+
+// computeParametersDelta compares two Parameter arrays and returns the new
+// parameters to add, to update and the parameter identifiers to delete
+func computeParametersDelta(
+	desired map[string]*string,
+	latest map[string]*string,
+) (map[string]*string, []string) {
+	toReset := []string{}
+	toModify := map[string]*string{}
+
+	for k, v := range desired {
+		if lv, found := latest[k]; !found {
+			toModify[k] = v
+		} else if !equalStrings(v, lv) {
+			toModify[k] = v
+		}
+	}
+	for k := range latest {
+		if _, found := desired[k]; !found {
+			toReset = append(toReset, k)
+		}
+	}
+	return toModify, toReset
+}
+
+// sliceStringChunks splits a supplied slice of string pointers into multiple
+// slices of string pointers of a given size.
+func sliceStringChunks(
+	input []string,
+	chunkSize int,
+) [][]string {
+	var chunks [][]string
+	for {
+		if len(input) == 0 {
+			break
+		}
+
+		if len(input) < chunkSize {
+			chunkSize = len(input)
+		}
+
+		chunks = append(chunks, input[0:chunkSize])
+		input = input[chunkSize:]
+	}
+
+	return chunks
+}
+
+// mapStringChunks splits a supplied map of string pointers into multiple
+// slices of maps of string pointers of a given size.
+func mapStringChunks(
+	input map[string]*string,
+	chunkSize int,
+) []map[string]*string {
+	var chunks []map[string]*string
+	chunk := map[string]*string{}
+	idx := 0
+	for k, v := range input {
+		if idx < chunkSize {
+			chunk[k] = v
+			idx++
+		} else {
+			// reset the chunker
+			chunks = append(chunks, chunk)
+			chunk = map[string]*string{}
+			idx = 0
+		}
+	}
+	chunks = append(chunks, chunk)
+
+	return chunks
+}
+
+type paramMeta struct {
+	isModifiable bool
+	isDynamic    bool
+}
+
+// metaFetcher is the functor we pass to the paramMetaCache that allows it to
+// fetch engine default parameter information
+type metaFetcher func(ctx context.Context, family string) (map[string]paramMeta, error)
+
+// paramMetaCache stores information about a parameter for a DB parameter group
+// family. We use this cached information to determine whether a parameter is
+// statically or dynamically defined (whether changes can be applied
+// immediately or pending a reboot) and whether a parameter is modifiable.
+//
+// Keeping things super simple for now and not adding any TTL or expiration
+// behaviour to the cache. Engine defaults are pretty static information...
+type paramMetaCache struct {
+	sync.RWMutex
+	hits   uint64
+	misses uint64
+	cache  map[string]map[string]paramMeta
+}
+
+// get retrieves the metadata for a named parameter group family and parameter
+// name.
+func (c *paramMetaCache) get(
+	ctx context.Context,
+	family string,
+	name string,
+	fetcher metaFetcher,
+) (*paramMeta, error) {
+	c.RLock()
+	defer c.RUnlock()
+
+	var err error
+	var found bool
+	var metas map[string]paramMeta
+	var meta paramMeta
+
+	metas, found = c.cache[family]
+	if !found {
+		c.misses++
+		metas, err = c.loadFamily(ctx, family, fetcher)
```

Comment:
> Not sure but this call might cause a deadlock (maybe just a timeout). Since `c.loadFamilly` also locks the mutex

Yeah, nice catch. I can manually call `RUnlock` after line 608 instead of `defer`'ing it